FEC-9807 Support Offline in KalturaPlayer (DTG)

x-NR-x opened the pull request:
Currently supporting only Basic and OTT Provider.

Solves FEC-9807

x-NR-x committed:
FEC-9807 Offline support WIP.

x-NR-x committed:
FEC-9807 Offline support WIP.

x-NR-x committed:
FEC-9807 Offline support for OTT provider.

x-NR-x committed:
FEC-9807 Added check for the startTime if it's nan.

x-NR-x committed:
FEC-10223 Changes to support the changes done in DTG.

noamtamim commented on Sources/Offline/OTT/OfflineManager+OTT.swift:
```diff
@@ -0,0 +1,84 @@
+                                                    partnerId: KalturaOTTPlayerManager.shared.partnerId,
+                                                    ks: mediaOptions.ks)
+        
+        // Add a function to the Utils
```
Please explain the comment.

noamtamim commented on Sources/Offline/OTT/OfflineManager+OTT.swift:
```diff
@@ -0,0 +1,84 @@
+        var referrer: String = "app://"
+        if let appId = Bundle.main.bundleIdentifier {
+            referrer += appId
+        } else {
+            PKLog.warning("The app's bundle identifier is not set")
+            referrer += "bundleIdentifier_is_empty"
+        }
```
How can it be empty?

noamtamim commented on Sources/Offline/OTT/OfflineManager+OTT.swift:
```diff
@@ -0,0 +1,84 @@
+        
+        // Add a function to the Utils
+        var referrer: String = "app://"
+        if let appId = Bundle.main.bundleIdentifier {
```
The app should be able to set the referrer, then it overrides this default.

x-NR-x replied:
Changed, It will be set in the setup func as an optional.
And now OfflineManager can fetch it from there.

noamtamim commented on Sources/Offline/AssetInfo.swift:
```diff
@@ -0,0 +1,84 @@
+    }
+    
+    public override var description: String {
+        return super.description + " itemId: \(itemId) state: \(state.description) estimatedSize: \(estimatedSize) downloadedSize: \(downloadedSize)"
```
Why super.description?

x-NR-x replied:
In order to print the object itself as well, for debugging purpose.

noamtamim commented on Sources/Offline/OfflineManager.swift:
```diff
@@ -0,0 +1,285 @@
+        
+        let itemId = mediaEntry.id
+        guard let mediaSource = localAssetsManager.getPreferredDownloadableMediaSource(for: mediaEntry) else {
+            PKLog.error("No preferred downloadable media source available.")
```
```suggestion
            PKLog.error("No downloadable media source available.")
```

noamtamim commented on Sources/Offline/OfflineManager.swift:
```diff
@@ -0,0 +1,285 @@
+    public func startAssetDownload(assetId: String) {
+        do {
+            try ContentManager.shared.startItem(id: assetId)
+        } catch {
+            PKLog.error(error.localizedDescription)
+        }
```
What about reporting the error to the app?
Same for other functions below. It seems that failure is ignored... ?

x-NR-x replied:
Fixed, please review the changes.

noamtamim commented on Sources/Offline/OfflineManager.swift:
```diff
@@ -0,0 +1,285 @@
+        return DRMStatus(fpsExpirationInfo)
+    }
+    
+    public func renewDrmAssetLicense(mediaEntry: PKMediaEntry) {
```
I think it should be "Asset DRM License".
Anyway, use either Drm or DRM in all places. For example, above you have `getDRMStatus?()`.

x-NR-x replied:
Fixed to renewAssetDRMLicense.

x-NR-x committed:
Update Sources/Offline/OfflineManager.swift

Updated message.

Co-authored-by: Noam Tamim <noam.tamim@kaltura.com>

x-NR-x committed:
FEC-10329 Fixed the code to use the referrer from the PlayKitUtils, removed it from the player options and added it to the setup of the KalturaPlayer in order to provide a custom value if required.

x-NR-x committed:
FEC-9807 Renamed renewDrmAssetLicense to renewAssetDRMLicense.
Added @objc in missing places.
Added throwing an error in some functions.

x-NR-x committed:
Merge branch 'FEC-10223' into FEC-9807

* FEC-10223:
  FEC-10223 Changes to support the changes done in DTG.

x-NR-x committed:
FEC-9807 Set version to the correct value, 4.0.0.
Set PlayKit to 3.17.
Added Dependance to KalturaPlayer/OTT inside Offline_OTT.
Added missing @objc.
Added a callback with an optional error for the renewAssetDRMLicense func.

x-NR-x committed:
FEC-9807 Fixed starting interrupted tasks when closing the app.

x-NR-x committed:
FEC-9807 Documentation added.
Added the missing userInfo data in the OfflineManagerError in case of a mediaProviderError.
Fixed the documentation in the KalturaOTTPlayer for the loadMedia function.

x-NR-x committed:
FEC-9807 Dependency to the KalturaPlayer/Core was missing in the Offline subspec.
